// socket.io-emitter/index.js

/**
 * Module dependencies.
 */

var client = require('redis').createClient;
var parser = require('socket.io-parser');
var msgpack = require('notepack.io');
var debug = require('debug')('socket.io-emitter');

/**
 * Module exports.
 */

module.exports = init;

/**
 * Flags.
 *
 * @api public
 */

var flags = [
  'json',
  'volatile',
  'broadcast'
];

/**
 * uid for emitter
 *
 * @api private
 */

var uid = 'emitter'; // 服务器标志

/**
 * Socket.IO redis based emitter.
 *
 * @param {Object} redis client (optional)
 * @param {Object} options
 * @api public
 */
// 用于外部应用给Socket.io系统发送消息
function init(redis, opts){
  opts = opts || {};

  if ('string' == typeof redis) {
    redis = client(redis);
  }

  if (redis && !redis.hset) { // 如果redis没有hset方法且是一个对象，则说明是配置
    opts = redis;
    redis = null;
  }

  if (!redis) { // 通过配置来初始化一个Redis客户端对象
    if (!opts.socket && !opts.host) throw new Error('Missing redis `host`');
    if (!opts.socket && !opts.port) throw new Error('Missing redis `port`');
    redis = opts.socket
      ? client(opts.socket)
      : client(opts.port, opts.host);
  }

  var prefix = opts.key || 'socket.io';

  return new Emitter(redis, prefix, '/'); // 这里只给默认命名空间发送消息
}

function Emitter(redis, prefix, nsp){
  this.redis = redis; // redis客户端对象
  this.prefix = prefix; // 前缀
  this.nsp = nsp; // 命名空间名称
  this.channel = this.prefix + '#' + nsp + '#'; // 命名空间全局频道名称

  this._rooms = [];
  this._flags = {};
}

/**
 * Apply flags from `Socket`.
 */

flags.forEach(function(flag){
  Emitter.prototype.__defineGetter__(flag, function(){
    debug('flag %s on', flag);
    this._flags[flag] = true;
    return this;
  });
});

/**
 * Limit emission to a certain `room`.
 *
 * @param {String} room
 */

Emitter.prototype.in =
Emitter.prototype.to = function(room){
  if (!~this._rooms.indexOf(room)) {
    debug('room %s', room);
    this._rooms.push(room);
  }
  return this;
};

/**
 * Return a new emitter for the given namespace.
 * 为给定的命名空间创建一个发射器
 *
 * @param {String} namespace
 */

Emitter.prototype.of = function(nsp){
  return new Emitter(this.redis, this.prefix, nsp);
};

/**
 * Send the packet.定义发射器的发射方法
 *
 * @api public
 */

Emitter.prototype.emit = function(){
  // packet
  var args = Array.prototype.slice.call(arguments);
  // 发送SocketIO协议的事件包
  var packet = { type: parser.EVENT, data: args, nsp: this.nsp };

  var opts = {
    rooms: this._rooms,
    flags: this._flags
  };

  // 对消息编码
  var msg = msgpack.encode([uid, packet, opts]);
  var channel = this.channel;
  // 如果是发送给指定的房间，则频道需要添加房间名称
  if (opts.rooms && opts.rooms.length === 1) {
    channel += opts.rooms[0] + '#';
  }
  debug('publishing message to channel %s', channel);
  this.redis.publish(channel, msg); // 发送频道消息

  // reset state
  this._rooms = [];
  this._flags = {};

  return this;
};
